src/app/modules/payments/stripe.utils.ts

```diff
@@ -0,0 +1,256 @@
+  `${
+    config.isDev ? 'http://localhost:5001' : config.app.appUrl
+  }/api/v3/payments/stripe/callback`
+
+/**
+ * Extracts the payment id from the metadata field of objects expected to have
+ * it (i.e. payment intents and charges).
+ * @param {Stripe.Metadata} metadata the metadata object which is expected to have a payment id
+ * @returns ok(paymentId) the extracted paymentId
+ * @returns err(StripeMetadataValidPaymentIdNotFoundError) if the payment id was not found or is an invalid BSON object id
+ */
+export const getMetadataPaymentId = (
+  metadata: Stripe.Metadata,
+): Result<string, StripeMetadataValidPaymentIdNotFoundError> => {
+  const logMeta = {
+    action: 'getMetadataPaymentId',
+    metadata,
+  }
+  const paymentId = get(metadata, 'paymentId') // TODO: Extract this value to a constant?
+  if (!paymentId || !mongoose.Types.ObjectId.isValid(paymentId)) {
+    logger.warn({
+      message: 'Got metadata with invalid paymentId',
+      meta: { ...logMeta, metadata },
+    })
+    return err(new StripeMetadataValidPaymentIdNotFoundError())
+  }
+  return ok(paymentId)
+}
+
+/**
+ * State machine that computes the state of the payment, given the list of
+ * Stripe events received via webhooks for this submission.
+ * @param {Stripe.Event[]} events the list of Stripe events for state to be computed on
+ * @returns ok({status, chargeIdLatest}) status and latest charge id based on the list of events
+ * @returns err(ComputePaymentStateError) if there is an error in computing the state
+ */
+export const computePaymentState = (
+  events: Stripe.Event[],
+): Result<
+  Pick<Payment, 'status' | 'chargeIdLatest'>,
+  ComputePaymentStateError
+> => {
+  const logMeta = {
+    action: 'computePaymentState',
+    events,
+  }
+
+  // Helper to get the charge id from a nested charge object.
+  const getChargeId = (charge: string | Stripe.Charge): string =>
+    typeof charge === 'string' ? charge : charge.id
+
+  // We only care about charge and dispute events for computing payment status.
+  // The event types are:
+  // - charge.(failed|pending|refunded|succeeded)
+  // - charge.dispute.(closed|created|funds_reinstated|funds_withdrawn|updated)
+  const chargeEvents = (events as Stripe.DiscriminatedEvent[]).filter(
+    (
+      event: Stripe.DiscriminatedEvent,
+    ): event is
+      | Stripe.DiscriminatedEvent.ChargeEvent
+      | Stripe.DiscriminatedEvent.ChargeDisputeEvent =>
+      event.type === 'charge.failed' ||
+      event.type === 'charge.pending' ||
+      event.type === 'charge.refunded' ||
+      event.type === 'charge.succeeded' ||
+      event.type.startsWith('charge.dispute.'),
+  )
+
+  // Step 1: State machine for computing the current value of the payment state.
+  // A "null" status represents an unknown state, and should be recovered from later.
+  let status: Payment['status'] | null = PaymentStatus.Pending
+  let chargeIdLatest: Payment['chargeIdLatest']
+
+  for (const event of chargeEvents) {
```

Comment:
Im thinking that we can extract this out as a reducer function, this will allow us to test it. 

```ts
  const chargeTransitionReducer = (
    state: ChargeStateType,
    curEvent:
      | Stripe.DiscriminatedEvent
      | Stripe.DiscriminatedEvent.ChargeDisputeEvent,
  ) => {
    switch (state.status) {
      case PaymentStatus.Pending:
      case PaymentStatus.Failed:
        if (curEvent.type === 'charge.failed') {
          return {
            status: PaymentStatus.Failed,
            chargeIdLatest: curEvent.data.object.id,
          }
        }
        if (curEvent.type === 'charge.succeeded') {
          return {
            status: PaymentStatus.Succeeded,
            chargeIdLatest: curEvent.data.object.id,
          }
        }
        return {
          status: null,
          chargeIdLatest: curEvent.data.object.id,
        }
       ....
  }


  const initialChargeState = { status: null, chargeIdLatest: null }
  type ChargeStateType = {
    status: Payment['status'] | null
    chargeIdLatest: Payment['chargeIdLatest'] | null
  }

  const chargeState: ChargeStateType = chargeEvents.reduce(
    applyChargeTransition,
    initialChargeState,
  )
  if (chargeState.status) return ok(chargeState)

```